### ISSUE TITLE
Proposal: Sink.log or alternative run...() method to log completion/failure

### ISSUE THREAD

**Issue by ktoso** [akka/akka#23501]
We often recommend using `runWith(Sink.ignore)` or `runWith(Sink.foreach)` the issue with those is that those will not log the failure. And people are left wondering "why the stream does not complete" or "why the stream does not run".

If they'd add a `.log()` they'd realise "oh, the filename has a typo" (had this exact case now).

Perhaps such "terminating" one should log the failure? 
Perhaps this should be configurable? `akka.stream.verbose-error-sinks = on` or something?

I propose to make it easier for beginners to notice when they made a mistake that made a failure reach a Sink, the hello-world even exposes the need for this:

```
object TestServer extends App {
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

FileIO.fromPath(new File("TYPO HERE").toPath)
    .via(Framing.delimiter(ByteString("\n"), maximumFrameLength = 1000))
    .map(_.utf8String)
    .map { line ⇒
      val entry = line.split(" ")
      val n = entry.head
      val time = entry.tail.head

      println(s"${n} ${Duration(time).toMicros}")
    } /*log ("progress")*/ runWith Sink.ignore
```

If I made a typo, I won't realise that I made a mistake. 

This of course is in-line with our error handling etc, however it's suprising to newcomers and we sometimes also get caught off guard by it (I just was, for a minute or so, but still).

Proposals (alternatives what we could do):

- Provide a Sink.log that by default would log completion and errors 
- Make our built-in "terminating, no outside communication" sinks like Sink.ignore log when the option `akka.stream.verbose-error-sinks = on` is on or something? This is a bit inconcistent, but could be pretty useful in debugging in any case, esp when an already live system has such issue and someone could not add the log() just like that.

**Comment by johanandren** [comment]
Where did the "always log stream failures" functionality go? I remember it was there for a while.

**Comment by patriknw** [comment]
`Sink.ignore` is not without outside communication. It returns a Future for this purpose.

I don't think a Sink.log is needed, one would not find that easier than adding `.log()`.

In another PR we discussed adding a logging attribute for additional logging. Perhaps that can also include logging in sinks such as `Sink.ignore`.

**Comment by ktoso** [comment]
> I don't think a Sink.log is needed, one would not find that easier than adding .log().

Yeah, that's true - and one would have to be aware of it.

I do like the config option a lot, we'd be "verbose" by default - log those errors etc. Let's find that other issue and weave such things into those sinks too :)

**Comment by johanandren** [comment]
The thing I was thinking of was logging any error failing a stage. Introduced by this PR: #19647

**Comment by johanandren** [comment]
One problem could be the not so good file connector API where the stream can complete successfully but with a `IOResult(_, Failed(ex))` though.

**Comment by richardimaoka** [comment]
I added a work-in-progress pull request as above. If anyone could advise on that, much appreciated.

By the way, after #21801 @ktoso's above code snippet actually logs the error without `log()`

```
[ERROR] [10/30/2017 05:28:14.246] [default-akka.stream.default-blocking-io-dispatcher-6] [akka://default/user/StreamSupervisor-0/flow-0-1-fileSource] Error during preStart in [FileSource(TYPO HERE, 8192)]: requirement failed: Path 'TYPO HERE' does not exist
java.lang.IllegalArgumentException: requirement failed: Path 'TYPO HERE' does not exist
        at scala.Predef$.require(Predef.scala:224)
        at akka.stream.impl.io.FileSource$$anon$1.preStart(IOSources.scala:73)
        at akka.stream.impl.fusing.GraphInterpreter.init(GraphInterpreter.scala:290)
        at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:540)
        at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:662)
        at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:710)
        at akka.actor.Actor$class.aroundPreStart(Actor.scala:528)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:653)
        at akka.actor.ActorCell.create(ActorCell.scala:591)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:462)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:484)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
        at akka.dispatch.Mailbox.run(Mailbox.scala:223)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

because now `FileSource` throws an exception from `preStart()`, 

https://github.com/akka/akka/blob/38622246d9ec82bfca78070b75beb93b06dc5a75/akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala#L82

which is captured by:
https://github.com/akka/akka/blob/38622246d9ec82bfca78070b75beb93b06dc5a75/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala#L293

We can still see this issue in a code like below.

```scala
Source(-5 to 5).map(1 / _)/*.log("")*/.runWith(Sink.ignore) // ArithmeticException: / by zero
```

No error is logged by that.

**Comment by ktoso** [comment]
Current direction I'd suggest to try is to make this a materializer setting that stages could look at perhaps, (simply an attribute, but also settable on the materializer)

**Comment by huntc** [comment]
Adding my 2 cents... I believe we should log all failures by default and have a configuration to disable such logging - similar to the actor deadletters behaviour. 

Earlier today, I got a service in production that, for some reason, has completed its UDP binding stream. I’m suspecting that it was fed some bad data and was unable to handle it. However, I can’t know for sure now... This is just one example. The lack of default failure logging has come up a few times over the years.

**Comment by patriknw** [comment]
Why isn’t there logging of failed Futures in standard library? I don’t think there is a default that fits all here. What we can do is to make it easier to turn on/off more logging.

**Comment by huntc** [comment]
> Why isn’t there logging of failed Futures in standard library?

That’s a great point. 

The only answer I can think of is that streams already have logging infrastructure and that it can be leveraged.
